design/20200305-spec-v1alpha2-dependency-parampassing.md

```diff
@@ -0,0 +1,96 @@
+# Component and Trait Dependency with Parameter Passing
+
+* Author: @wonderflow @hongchaodeng
+
+## Background
+
+In https://github.com/oam-dev/spec/issues/171, we have discussed our dependency stories and started to explore solutions.
+Now that we have designed and implemented mechanism to satisfy dependency requirements, we are proposing to add them into OAM spec.
+
+## Use Cases
+
+### Case 1: MySQL and Web
+
+Two components, MySQL and Web exists in the same AppConfig and Web depends on MySQL that:
+
+1. MySQL needs to start first;
+2. MySQL connection info needs to be injected into Web's environment variables or files;
+
+### Case 2: NAS FileSystem and MountTarget
+
+Two components, NAS_FileSystem and NAS_MountTarget exists in the same AppConfig and NAS_MountTarget depends on NAS_FileSystem that:
+
+1. NAS_FileSystem needs to be created first;
+2. NAS_MountTarget needs filesystemID from NAS_FileSystem to fill in NAS_MountTarget's spec.
+
+## Proposal
+
+### Component and Trait Dependency
+
+We are proposing to add `dependencies` in AppConfig as follows:
+
+```yaml
+kind: ApplicationConfiguration
+spec:
+- components:
+  - componentName: mysql
+  - componentName: web
+    dependencies:
+      - mysql
+```
+
+Here a component can describe its dependencies, or dependent components.
+
+For Case 1, we also suggest to use ServiceBinding in [trait-injector](https://github.com/oam-dev/trait-injector) to inject secret. It is a trait that should be applied before component starts. We propose to add `ordering` to express trait's ordering w.r.t the component.
+
+```yaml
+- componentName: web
+  dependencies:
+    - mysql
+  traits:
+    - ordering: preStart # applied before component starts
+      trait:
+        kind: ServiceBinding
+        spec:
+          bindings:
+          - from:
+              secret:
+                name: mysql-conn
+            to: # inject into Web Deployment's file path
+              filePath: /input/mysql-conn
+```
+
+### Parameter Passing
```

Comment:
Is there any reason why we can't use serviceBinding trait here too?

Reply:
I have explained why these two cases are different. So parameter passing here is more likely to be used as internal arguments delivery, while service binding aims to be used in a more common way.

Reply:
More than that, we don't want to make service binding be coupled with parameters. Parameters are invented for application developers to tell what they need to application operators, while app operator can override the parameter. We could take parameter passing as an advanced override, it allows app operator override parameter from component output. 

In this view, they have nothing to do with service binding, right?

Reply:
If I follow correctly, the problem we're solving here is that when workload A is created by appconfig X it may want to depend on some property of workload B, which is also created by appconfig X. It's possible that the property workload A depends on will be sensitive (e.g. a shared secret).

It seems like we're proposing two ways of addressing this:

1. The `ServiceBinding` trait, which requests that the data from a named secret (which is presumably produced by worklad B) be mounted at a named path within workload A.
1. Parameter passing, which requests that a parameter of workload A be set to the value of a fieldpath within workload B.

It seems like parameter passing is simpler, but `ServiceBinding` is more flexible. Would OAM be simpler overall if we supported only `ServiceBinding` and not parameter passing?

Reply:
`parameter passing` is also required as it gives us a way to pass through parameters directly between components. By using `ServiceBinding` trait, we are able to pass through `object` level information between components, while `parameter passing` can support `spec field` level information.

Reply:
Could you help me understand what the difference between object level and spec field level information?

Reply:
For example: 

Assuming below is an existing CRD + Operator which can receive filesystemID as part of spec in CR.

```
   apiVersion: ros.aliyun.com/v1alpha2
   kind: NASMountTarget
   spec:
      filesystemID: "" # to be filled
```
If we give a secret object to it, it has no way to consume this secret.

While if the workload is K8s deployment, it has the way to consume secret object:

```
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
        envFrom:
        - secretRef:
            name: my-secret
```